src/System.Console/src/System/IO/StdInStreamReader.cs

```diff
@@ -0,0 +1,350 @@
+                    {
+                        isShift = true;
+                        return ConsoleKey.A + (x - 'A');
+                    }
+
+                    // 4. a to z.
+                    if (x >= 'a' && x <= 'z')
+                    {
+                        return ConsoleKey.A + (x - 'a');
+                    }
+
+                    break;
+            }
+
+            return default(ConsoleKey);
+        }
+
+        internal bool TryMapBufferToConsoleKey(out ConsoleKey key, out char ch, out bool isShift, out bool isAlt, out bool isCtrl)
+        {
+            Debug.Assert(!IsExtraBufferEmpty());
+
+            isAlt = isCtrl = isShift = false;
+            int keyLength;
+
+            // 1. Try to get the special key match from the TermInfo static information.
+            if (ConsolePal.TryGetSpecialConsoleKey(unprocessedBufferToBeRead, startIndex, endIndex, out key, out keyLength))
+            {
+                startIndex += keyLength;
+                ch = '\0'; //Special keys are non-printable.
+                return true;
+            }
+
+            // 1. Check if we can match Esc + combination and guess if alt was pressed.
+            if (isAlt == false && unprocessedBufferToBeRead[startIndex] == (char)(int)(0x1B) /*Alt is send as an escape character*/ && endIndex - startIndex + 1 >= 2 /*We have at least two characters to read.*/)
+            {
+                isAlt = true; // Since the alt is pressed.
+                startIndex++;
+                if (TryMapBufferToConsoleKey(out key, out ch, out isShift, out isAlt, out isCtrl))
+                {
+                    isAlt = true;
+                    return true;
+                }
+                else
+                {
+                    // We could not find a matching key here so, Alt+ combination assumption is in-correct.
+                    // The current key needs to be marked as Esc key.
+                    // Also, we do not increment startIndex as we already did it.
+                    key = ConsoleKey.Escape;
+                    ch = (char)(int)(0x1B);
+                    isAlt = false;
+                    return true;
+                }
+            }
+
+            // Try reading the first char in the buffer and interpret it as a key.
+            ch = unprocessedBufferToBeRead[startIndex++];
+            key = GetKeyFromCharValue(ref ch, out isShift, out isCtrl);
+            return key != (ConsoleKey)0;
+
+        }
+
+        internal ConsoleKeyInfo ReadKeyInternal()
+        {
+            ConsoleKey key;
+            char ch;
+            bool isAlt, isCtrl, isShift;
+
+            if (IsExtraBufferEmpty())
+            {
+                ReadExtraBuffer();
+            }
+
+            if(TryMapBufferToConsoleKey(out key, out ch, out isShift, out isAlt, out isCtrl))
+            {
+                return new ConsoleKeyInfo(ch, key, isShift, isAlt, isCtrl);
+            }
+
+            return default(ConsoleKeyInfo);
```

Comment:
On windows for every key we always get a virtual key code, so even if we do not have a value for it in ConsoleKey we simply return the virtual key code. Since, adding a new state to the ConsoleKey enum would require update to the other platforms as well, I will do as you suggested use  default(ConsoleKey).

In fact since now we have a way to not lose information, I will remove ConsoleKey.OEM* keys from Unix implementation as well i.e. all the special characters like !,@,# because there is no real guarantee that they will always map to the same OEM key and the current implementation was best effort.